staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -300,7 +305,24 @@ func (w *watchCache) processEvent(event watch.Event, resourceVersion uint64, upd
 }
 
 // Assumes that lock is already held for write.
+// Use dynamic size cache. If all cached events occur inside eventFreshDuration, increase cache capacity by 2x.
+// Decrease cache capacity by 2x when recent half events occur outside eventFreshDuration.
 func (w *watchCache) updateCache(event *watchCacheEvent) {
+	if w.endIndex == w.startIndex+w.capacity && event.RecordTime.Sub(w.cache[w.startIndex%w.capacity].RecordTime) < eventFreshDuration {
+		w.makeNewCache(w.capacity * 2)
+		w.capacity *= 2
+	} else if w.endIndex-w.startIndex < w.capacity/2 {
```

Comment:
I'm not sure we even need this condition. This can only be smaller than capacity before we process the first events - starting from then we will always have the cache full, because we remove a single item only when we need a place to insert a new one. I would remove it.

Reply:
removed